### PR TITLE
SAAS-3794: Added App Mesh Restrict External Traffic Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: plugins/aws/appmesh/restrictExternalTraffic.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+module.exports = {
+    title: 'App Mesh Restrict External Traffic',
+    category: 'App Mesh',
+    domain: 'Content Delivery',
+    description: 'Ensure that virtual nodes have egress only access to other defined resources available within the service mesh.',
+    more_info: 'Amazon App Mesh gives control over the traffic flow to configure the microservices. Getting the control will allow you to decide if you want to give access to outside world for communication or not.',
+    link: 'https://docs.aws.amazon.com/app-mesh/latest/userguide/meshes.html',
+    recommended_action: 'Enable the feature to allow egress only from virtual nodes to other defined resources in the service mesh.',
+    apis: ['AppMesh:listMeshes', 'AppMesh:describeMesh'],
+
```
Comment (m-nouman-suleman) [inline plugins/aws/appmesh/restrictExternalTraffic.js:13]:
```suggestion
```

File: plugins/aws/appmesh/restrictExternalTraffic.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                var describeMesh = helpers.addSource(cache, source,
+                    ['appmesh', 'describeMesh', region, mesh.meshName]);
+
+
+                if (!describeMesh || describeMesh.err || !describeMesh.data ||
+                    !describeMesh.data.mesh) {
+                    helpers.addResult(results, 3,
+                        `Unable to get App Mesh description: ${helpers.addError(describeMesh)}`,
+                        region, resource);
+                    continue;
```
Comment (m-nouman-suleman) [inline plugins/aws/appmesh/restrictExternalTraffic.js:51]:
```suggestion
                    continue;
                    rcb();
```

File: plugins/aws/appmesh/restrictExternalTraffic.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                } 
+
+                if (describeMesh.data.mesh.spec &&
+                    describeMesh.data.mesh.spec.egressFilter &&
+                    describeMesh.data.mesh.spec.egressFilter.type.toUpperCase() === 'DROP_ALL') {
+                    helpers.addResult(results, 0,
+                        'Amazon App Mesh allows egress only from virtual nodes to other defined resources.',
+                        region, resource);
+                    continue;
```
Comment (m-nouman-suleman) [inline plugins/aws/appmesh/restrictExternalTraffic.js:60]:
```suggestion
                    continue;
                    rcb();
```

File: plugins/aws/appmesh/restrictExternalTraffic.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                    describeMesh.data.mesh.spec.egressFilter.type.toUpperCase() === 'DROP_ALL') {
+                    helpers.addResult(results, 0,
+                        'Amazon App Mesh allows egress only from virtual nodes to other defined resources.',
+                        region, resource);
+                    continue;         
+                } else {
+                    helpers.addResult(results, 2,
+                        'Amazon App Mesh allows egress to any endpoint inside or outside of the service mesh.',
+                        region, resource);
+                    continue;
```
Comment (m-nouman-suleman) [inline plugins/aws/appmesh/restrictExternalTraffic.js:65]:
```suggestion
                    rcb();
                    
```

File: plugins/aws/appmesh/restrictExternalTraffic.spec.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+            }
+        },
+    };
+};
+
+describe('restrictExternalTraffic', function () {
+    describe('run', function () {
+        it('should PASS if App Mesh allows egress only from virtual nodes to other defined resources.', function (done) {
+            const cache = createCache([listMeshes[0]], describeMesh[1]);
+            restrictExternalTraffic.run(cache, { }, (err, results) => {
```
Comment (m-nouman-suleman) [inline plugins/aws/appmesh/restrictExternalTraffic.spec.js:90]:
```suggestion
            restrictExternalTraffic.run(cache, {}, (err, results) => {
```

File: plugins/aws/appmesh/restrictExternalTraffic.js (hunk 4 above)
Comment (m-nouman-suleman) [reply plugins/aws/appmesh/restrictExternalTraffic.js:65]: @AkhtarAmir bhai please verify if this should be changed as per my knowledge it should be

Comment (sadeed12345) [reply plugins/aws/appmesh/restrictExternalTraffic.js:65]: @m-nouman-suleman bhai isn't it inside the for loop, so shouldn't they just continue?

File: plugins/aws/appmesh/restrictExternalTraffic.spec.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+            restrictExternalTraffic.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(3);
+                expect(results[0].region).to.equal('us-east-1');
+                done();
+            });
+        });
+
+        it('should UNKNOWN if Unable to get App Mesh description', function (done) {
+            const cache = createCache([listMeshes[0]], null, { message: "Unable to get App Mesh description" });
```
Comment (sadeed12345) [inline plugins/aws/appmesh/restrictExternalTraffic.spec.js:132]:
```suggestion
            const cache = createCache([listMeshes[0]], null, null, { message: "Unable to get App Mesh description" });
```

File: plugins/aws/appmesh/restrictExternalTraffic.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'App Mesh Restrict External Traffic',
+    category: 'App Mesh',
+    domain: 'Content Delivery',
+    description: 'Ensure that virtual nodes have egress only access to other defined resources available within the service mesh.',
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/restrictExternalTraffic.js:8]:
```suggestion
    description: 'Ensure that Amazon App Mesh virtual nodes have egress only access to other defined resources available within the service mesh.',
```

File: plugins/aws/appmesh/restrictExternalTraffic.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+        async.each(regions.appmesh, function(region, rcb){        
+            var listMeshes = helpers.addSource(cache, source,
+                ['appmesh', 'listMeshes', region]);
+
+            if (!listMeshes) return rcb();
+
+            if (listMeshes.err || !listMeshes.data) {
+                helpers.addResult(results, 3,
+                    `Unable to query for App Meshes: ${helpers.addError(listMeshes)}`,region);
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/restrictExternalTraffic.js:27]:
```suggestion
                    `Unable to query for App Mesh meshes: ${helpers.addError(listMeshes)}`,region);
```

File: plugins/aws/appmesh/restrictExternalTraffic.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+                return rcb();
+            }
+
+            if (!listMeshes.data.length) {
+                helpers.addResult(results, 0, 'No App Meshes found', region);
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/restrictExternalTraffic.js:32]:
```suggestion
                helpers.addResult(results, 0, 'No App Mesh meshes found', region);
```

File: plugins/aws/appmesh/restrictExternalTraffic.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+            }
+
+            for (let mesh of listMeshes.data) {
+                if (!mesh.arn) continue;
+
+                let resource = mesh.arn;
+
+                var describeMesh = helpers.addSource(cache, source,
+                    ['appmesh', 'describeMesh', region, mesh.meshName]);
+
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/restrictExternalTraffic.js:43]:
```suggestion
```

File: plugins/aws/appmesh/restrictExternalTraffic.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+                        `Unable to get App Mesh description: ${helpers.addError(describeMesh)}`,
+                        region, resource);
+                    continue;
+                } 
+
+                if (describeMesh.data.mesh.spec &&
+                    describeMesh.data.mesh.spec.egressFilter &&
+                    describeMesh.data.mesh.spec.egressFilter.type.toUpperCase() === 'DROP_ALL') {
+                    helpers.addResult(results, 0,
+                        'Amazon App Mesh allows egress only from virtual nodes to other defined resources.',
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/restrictExternalTraffic.js:57]:
```suggestion
                        'Amazon App Mesh allows egress only from virtual nodes to other defined resources',
```

File: plugins/aws/appmesh/restrictExternalTraffic.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+                        region, resource);
+                    continue;         
+                } else {
+                    helpers.addResult(results, 2,
+                        'Amazon App Mesh allows egress to any endpoint inside or outside of the service mesh.',
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/restrictExternalTraffic.js:62]:
```suggestion
                        'App Mesh mesh allows access to all endpoint inside or outside of the service mesh.',
```

File: plugins/aws/appmesh/restrictExternalTraffic.js [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+                        `Unable to describe App Mesh mesh: ${helpers.addError(describeMesh)}`,
+                        region, resource);
+                    continue;
+                } 
+
+                if (describeMesh.data.mesh.spec &&
+                    describeMesh.data.mesh.spec.egressFilter &&
+                    describeMesh.data.mesh.spec.egressFilter.type.toUpperCase() === 'ALLOW_ALL') {
+                    helpers.addResult(results, 2,
+                        'App Mesh mesh allows allows all external egress traffic',
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/restrictExternalTraffic.js:58]:
```suggestion
                        'App Mesh mesh allows access to external services',
```

File: plugins/aws/appmesh/restrictExternalTraffic.js [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+                        region, resource);       
+                } else {
+                    helpers.addResult(results, 0,
+                        'App Mesh mesh denies all external egress traffic',
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/restrictExternalTraffic.js:62]:
```suggestion
                        'App Mesh mesh does not allow access to external services',
```

File: plugins/aws/appmesh/restrictExternalTraffic.spec.js [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+};
+
+describe('restrictExternalTraffic', function () {
+    describe('run', function () {
+        it('should PASS if App Mesh mesh denies all external egress traffic', function (done) {
+            const cache = createCache([listMeshes[0]], describeMesh[1]);
+            restrictExternalTraffic.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('App Mesh mesh denies all external egress traffic');
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/restrictExternalTraffic.spec.js:93]:
```suggestion
                expect(results[0].message).to.include('App Mesh mesh does not allow access to external services');
```

File: plugins/aws/appmesh/restrictExternalTraffic.spec.js [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+                done();
+            });
+        });
+
+        it('should FAIL if App Mesh mesh allows allows all external egress traffic', function (done) {
+            const cache = createCache([listMeshes[0]], describeMesh[0]);
+            restrictExternalTraffic.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
+                expect(results[0].message).to.include('App Mesh mesh allows allows all external egress traffic');
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/restrictExternalTraffic.spec.js:104]:
```suggestion
                expect(results[0].message).to.include(App Mesh mesh allows access to external services');
```

File: plugins/aws/appmesh/restrictExternalTraffic.spec.js [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+            restrictExternalTraffic.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('App Mesh mesh denies all external egress traffic');
+                expect(results[0].region).to.equal('us-east-1');
+                done();
+            });
+        });
+
+        it('should FAIL if App Mesh mesh allows allows all external egress traffic', function (done) {
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/restrictExternalTraffic.spec.js:99]:
```suggestion
        it('should FAIL if App Mesh mesh allows access to external services', function (done) {
```

File: plugins/aws/appmesh/restrictExternalTraffic.spec.js [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+                    }
+                }
+            }
+        },
+    };
+};
+
+describe('restrictExternalTraffic', function () {
+    describe('run', function () {
+        it('should PASS if App Mesh mesh denies all external egress traffic', function (done) {
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/restrictExternalTraffic.spec.js:88]:
```suggestion
        it('should PASS if App Mesh mesh does not allow access to external services', function (done) {
```

File: collectors/aws/collector.js [hunk 19]
```diff
@@ -1108,6 +1108,14 @@ var postcalls = [
...
             }
         },
+        AppMesh: {
+            describeMesh: {
+                reliesOnService: 'appmesh',
+                reliesOnCall: 'listMeshes',
+                filterKey: 'meshName',
+                filterValue: 'meshName'
+            }
+        },
```
Comment (AkhtarAmir) [inline collectors/aws/collector.js:1118]:
```suggestion
```

File: collectors/aws/collector.js [hunk 20]
```diff
@@ -1098,6 +1098,12 @@ var postcalls = [
                 reliesOnCall: 'listMeshes',
                 filterKey: 'meshName',
                 filterValue: 'meshName'
+            },
+            describeMesh: {
+                reliesOnService: 'appmesh',
+                reliesOnCall: 'listMeshes',
+                filterKey: 'meshName',
+                filterValue: 'meshName'
```
Comment (AkhtarAmir) [inline collectors/aws/collector.js:1106]:
```suggestion
```

**Pull request merged**